source/MRViewer/MRObjectTransformWidget.cpp

```diff
@@ -93,6 +96,27 @@ float findAngleDegOfPick( const Vector3f& center, const Vector3f& zeroPoint, con
     return angleRes;
 }
 
+void decomposeQR( const Matrix3f& m, Matrix3f& q, Matrix3f& r )
+{
+    Eigen::HouseholderQR<Eigen::MatrixXf> qr( toEigen( m ) );
+    q = fromEigen( Eigen::Matrix3f{ qr.householderQ() } );
+    r = fromEigen( Eigen::Matrix3f{ qr.matrixQR() } );
+    r.y.x = r.z.x = r.z.y = 0;
+}
+
+void decomposePositiveQR( const Matrix3f& m, Matrix3f& q, Matrix3f& r )
+{
+    decomposeQR( m, q, r );
+    Matrix3f sign;
+    for ( int i = 0; i < 3; ++i )
+    {
+        if ( r[i][i] < 0 )
+            sign[i][i] = -1;
+    }
+    q = q * sign;
+    r = sign * r;
+}
```

Comment:
we have same funcitons in ImGuiMenu, mb make separate functoin like `extractScale( AffineXf3f& xf )` to separate scale from rigid transform somewhere in MRMesh project?

Reply:
Implemented.